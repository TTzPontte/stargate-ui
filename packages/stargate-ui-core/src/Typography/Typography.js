import React from 'react';
import PropTypes from 'prop-types';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import Factory from '../Factory';
import Svg from '../Svg';

const TYPOGRAPHY_VARIANT_H1 = 'h1';
const TYPOGRAPHY_VARIANT_H2 = 'h2';
const TYPOGRAPHY_VARIANT_H3 = 'h3';
const TYPOGRAPHY_VARIANT_H4 = 'h4';
const TYPOGRAPHY_VARIANT_H5 = 'h5';
const TYPOGRAPHY_VARIANT_H6 = 'h6';
const TYPOGRAPHY_VARIANT_BODY = 'body';
const TYPOGRAPHY_VARIANT_BODYLOWER = 'bodylower';
const TYPOGRAPHY_VARIANT_SMALL = 'small';
const TYPOGRAPHY_VARIANT_TINY = 'tiny';

const TYPOGRAPHY_VARIANT = {
  [TYPOGRAPHY_VARIANT_H1]: 'h1',
  [TYPOGRAPHY_VARIANT_H2]: 'h2',
  [TYPOGRAPHY_VARIANT_H3]: 'h3',
  [TYPOGRAPHY_VARIANT_H4]: 'h4',
  [TYPOGRAPHY_VARIANT_H5]: 'h5',
  [TYPOGRAPHY_VARIANT_H6]: 'h6',
  [TYPOGRAPHY_VARIANT_BODY]: 'p',
  /**
   * @todo needs improvement
   */
  // [TYPOGRAPHY_VARIANT_BODYLOWER]: 'p',
  [TYPOGRAPHY_VARIANT_SMALL]: 'span',
  [TYPOGRAPHY_VARIANT_TINY]: 'small',
};

const styles = (theme) => {
  const {
    breakpoints,
    unit,
    maxWidth,
    typography: {
      fontFamily,
      fontSize,
      lineHeight,
      fontWeight,
    },
  } = theme;

  /**
   * @todo add support for custom color
   */
  const typography = {
    margin: 0,
    letterSpacing: .5,
    fontFamily: ({ variant }) => (
      theme.typography?.[variant].fontFamily || fontFamily
    ),
    fontSize: ({ variant }) => {
      theme.typography?.[variant].fontSize || fontSize
    },
    lineHeight: ({ variant }) => (
      theme.typography?.[variant].lineHeight || lineHeight
    ),
    fontWeight: ({ variant, weight }) => (
      weight || theme.typography?.[variant].fontWeight || fontWeight
    ),
    textTransform: ({ transform }) => (
      transform
    ),
    display: (props) => {
      const {
        variant,
        paragraph,
        display,
      } = props;

      return display || ((paragraph || [
        TYPOGRAPHY_VARIANT_H1,
        TYPOGRAPHY_VARIANT_H2,
        TYPOGRAPHY_VARIANT_H3,
        TYPOGRAPHY_VARIANT_H4,
        TYPOGRAPHY_VARIANT_H5,
        TYPOGRAPHY_VARIANT_H6,
        TYPOGRAPHY_VARIANT_BODY,
      ].includes(variant)) && 'block') || 'initial';
    },
  };

  const typographyQuote = {
    position: 'relative',
    zIndex: 2,
  };

  const typographyQuoteIcon = {
    position: 'absolute',
    opacity: .2,
    top: 0,
    left: 0,
    zIndex: 1,
    width: ({ variant }) => (
      (theme.typography?.[variant].fontSize || fontSize) * 4
    ),
  };

  const typographyFluid = ({ variant }) => {
    const fontVariant = theme.typography?.[variant];
    const { max, min } = fontVariant?.fontSizeFluid;
    const fontSize = unit.fluid(max, min, maxWidth);

    return {
      fontSize,
      [breakpoints.up('lg')]: {
        fontSize: theme.typography?.[variant].fontSize || fontSize,
      },
    }
  };

  return {
    typography,
    typographyQuote,
    typographyQuoteIcon,
    typographyFluid,
  };
};

const Typography = (props) => {
  const {
    color,
    display,
    transform,
    paragraph,
    gutter,
    quote,
    children,
    weight,
    fluid = true,
    variant: typographyVariant,
    element: elementTagString = 'p',
    className: inheritedClassName,
    ...factoryProps
  } = props;

  const typographyMapping = TYPOGRAPHY_VARIANT;
  const element = elementTagString || typographyMapping[typographyVariant];
  /**
   * @todo need to filter equal elements
   */
  const variant = typographyVariant || Object.keys(typographyMapping).reduce(
    (a, b) => typographyMapping[b] === elementTagString ? b : a
  );

  const [
    {
      typography: classTypography,
      typographyQuote: classTypographyQuote,
      typographyQuoteIcon: classTypographyQuoteIcon,
      typographyFluid: classTypographyFluid,
    },
  ] = useStyles(styles, {
    variant,
    transform,
    color,
    paragraph,
    gutter,
    display,
    weight,
  });
  const className = clsx(
    classTypography,
    {
      [classTypographyQuote]: quote,
      [classTypographyFluid]: fluid,
    },
    inheritedClassName,
  );

  const marginBottom = ((gutter || paragraph || [
    TYPOGRAPHY_VARIANT_H1,
    TYPOGRAPHY_VARIANT_H2,
    TYPOGRAPHY_VARIANT_H3,
    TYPOGRAPHY_VARIANT_H4,
    TYPOGRAPHY_VARIANT_H5,
    TYPOGRAPHY_VARIANT_H6,
    TYPOGRAPHY_VARIANT_BODY,
  ].includes(variant)) && (gutter ?? 2)) || 0;

  return (
    <Factory
      element={element}
      className={className}
      marginBottom={marginBottom}
      color={color}
      {...factoryProps}
    >
      {quote && (
        <Svg className={classTypographyQuoteIcon}>
          {/* @todo path does not fit */}
          <svg>
            <path fill="currentColor" transform="translate(-995.205 -1888.884)" d="M1004.8,1910.5h-9.6v-9.5c0-5.7,3.2-8.9,9.6-9.6v4.8
              c-3.1,0.1-4.7,1.7-4.8,4.8h4.8V1910.5z M1019.2,1910.5h-9.6v-9.5c0-5.7,3.2-8.9,9.6-9.6v4.8c-3.1,0.1-4.7,1.7-4.8,4.8h4.8V1910.5z"
            />
          </svg>
        </Svg>
      )}
      {children}
    </Factory>
  );
};

Typography.propTypes = {
  children: PropTypes.node.isRequired,
  /**
   * @default p
   */
  element: PropTypes.elementType,
  /**
   * @default TYPOGRAPHY_VARIANT_BODY
   */
  variant: PropTypes.oneOf([
    TYPOGRAPHY_VARIANT_H1,
    TYPOGRAPHY_VARIANT_H2,
    TYPOGRAPHY_VARIANT_H3,
    TYPOGRAPHY_VARIANT_H4,
    TYPOGRAPHY_VARIANT_H5,
    TYPOGRAPHY_VARIANT_H6,
    TYPOGRAPHY_VARIANT_BODY,
    TYPOGRAPHY_VARIANT_BODYLOWER,
    TYPOGRAPHY_VARIANT_SMALL,
    TYPOGRAPHY_VARIANT_TINY,
  ]),
  /**
   * @default inherit
   */
  transform: PropTypes.oneOf([
    'inherit',
    'initial',
    'uppercase',
    'lowercase',
    'capitalize',
  ]),
  /**
   * @default inherit
   */
  color: PropTypes.oneOf([
    'inherit',
    'initial',
    'primary',
    'secondary',
    'success',
    'warning',
    'error',
    'info',
  ]),
  paragraph: PropTypes.bool,
  /**
   * @default 2
   */
  gutter: PropTypes.number,
  /**
   * @default initial
   */
  display: PropTypes.oneOf([
    'initial',
    'inherit',
    'block',
    'inline',
  ]),
};

export default Typography;
