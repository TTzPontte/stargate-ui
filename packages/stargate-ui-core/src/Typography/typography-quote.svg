<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 24.0.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 width="24px" height="24px" viewBox="0 0 24 24" style="enable-background:new 0 0 24 24;" xml:space="preserve">
<path id="Aspas" transform="translate(-995.205 -1888.884)" d="M1004.8,1910.5h-9.6v-9.5c0-5.7,3.2-8.9,9.6-9.6v4.8
	c-3.1,0.1-4.7,1.7-4.8,4.8h4.8V1910.5z M1019.2,1910.5h-9.6v-9.5c0-5.7,3.2-8.9,9.6-9.6v4.8c-3.1,0.1-4.7,1.7-4.8,4.8h4.8V1910.5z"
	/>
</svg>
