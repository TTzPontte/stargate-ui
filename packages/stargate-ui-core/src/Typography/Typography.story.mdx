import {
  Meta,
  Canvas,
  Status,
} from '@pontte/stargate-ui-playground';
import Typography from './Typography';

export const TypographyStory = (props) => (
  <Typography {...props}>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Beatae cum quidem officiis dolor impedit corrupti recusandae asperiores nesciunt architecto ea hic, error dolores fugiat. Eius itaque iure neque beatae ducimus.
  </Typography>
);

<Meta title="Core/Typography" component={Typography} />

<Story
  name="Explore Properties"
  parameters={{
    docs: { disable: true },
    jest: ['Typography.spec.jsx'],
  }}
>
  {TypographyStory.bind({})}
</Story>

## Typography

<Status type="stable" />

### General

Stargate UI uses *Lato* as font default but **not includes font files** automatically.

### Install

You can install from [typeface-lato](https://www.npmjs.com/package/typeface-lato).

```sh
npm install typace-lato -S
```

or

```sh
yarn add typace-lato
```

Then, import it in your entry-point.

```jsx
import 'typace-lato';
```

### Google Fonts

If you want saving some bits from your server, you can add from [Google Fonts Lato](https://fonts.google.com/specimen/Lato).

```html
<link href="https://fonts.googleapis.com/css2?family=Lato:ital,wght@0,100;0,300;0,400;0,700;0,900;1,100;1,300;1,400;1,700;1,900&display=swap" rel="stylesheet">
```

### Heading

<Canvas>
  <div>
    <Typography element="h1">h1. Heading</Typography>
    <Typography element="h2">h2. Heading</Typography>
    <Typography element="h3">h3. Heading</Typography>
    <Typography element="h4">h4. Heading</Typography>
    <Typography element="h5">h5. Heading</Typography>
    <Typography element="h6">h6. Heading</Typography>
  </div>
</Canvas>

### Body Variant

<Canvas>
  <Typography variant="body" paragraph>
    <b>body.</b> Lorem ipsum dolor sit amet, consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
    veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
    consequat. Duis aute irure dolor in reprehenderit in voluptate
    velit esse cillum dolore eu fugiat nulla pariatur
  </Typography>
  <Typography variant="bodylower" paragraph>
    <b>bodylower.</b> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
    nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
    velit esse cillum dolore eu fugiat nulla pariatur
  </Typography>
  <Typography variant="small" paragraph>
    <b>small.</b> Lorem ipsum dolor sit amet, consectetur adipisicing elit,
    sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
    veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
    consequat. Duis aute irure dolor in reprehenderit in voluptate
    velit esse cillum dolore eu fugiat nulla pariatur
  </Typography>
  <Typography variant="tiny" paragraph>
    <b>tiny.</b> Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
    nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
    velit esse cillum dolore eu fugiat nulla pariatur
  </Typography>
</Canvas>


### Colors

<Canvas>
  <Typography variant="body" color="primary">
    Hello world in <i>primary</i> color!
  </Typography>
  <Typography variant="body" color="secondary">
    Hello world in <i>secondary</i> color!
  </Typography>
  <Typography variant="body" color="success">
    Hello world in <i>success</i> color!
  </Typography>
  <Typography variant="body" color="warning">
    Hello world in <i>warning</i> color!
  </Typography>
  <Typography variant="body" color="error">
    Hello world in <i>error</i> color!
  </Typography>
  <Typography variant="body" color="info">
    Hello world in <i>info</i> color!
  </Typography>
</Canvas>


### Display

<Canvas>
  <Typography element="span" display="block">
    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
    nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
    velit esse cillum dolore eu fugiat nulla pariatur
  </Typography>
</Canvas>


### Alignment

Typography uses Factory as base, then accepts each its properties.

<Canvas>
  <div>
    <Typography gutter={0}>
        Align to <b>left</b>
    </Typography>
    <Typography gutter={0} textAlign="center">
        Align to <b>center</b>
    </Typography>
    <Typography gutter={4} textAlign="right">
        Align to <b>center</b>
    </Typography>
    <Typography textAlign="justify">
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut
      labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
      nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur
    </Typography>
  </div>
</Canvas>
