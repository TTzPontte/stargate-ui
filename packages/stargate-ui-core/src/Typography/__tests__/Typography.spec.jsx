import React from 'react';

import Typography from '../Typography';

describe('Typography', () => {
  it('should render properly', () => {
    const { container } = render(
      <Typography>
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minima reiciendis id nemo dicta! Soluta voluptate cupiditate corrupti architecto rerum nisi excepturi sunt quaerat iste quibusdam, non nam possimus ut eveniet.
      </Typography>,
    );

    expect(container).toMatchSnapshot();
  });
});
