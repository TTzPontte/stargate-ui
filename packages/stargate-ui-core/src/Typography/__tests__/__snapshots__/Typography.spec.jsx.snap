// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography should render properly 1`] = `
<div>
  <p
    class="stargate-ui-factory-0-2-8 stargate-ui-factory-d0-0-2-10 stargate-ui-typography-0-2-1 stargate-ui-typography-d0-0-2-5 stargate-ui-typographyFluid-0-2-4 stargate-ui-typographyFluid-d2-0-2-7"
  >
    Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minima reiciendis id nemo dicta! Soluta voluptate cupiditate corrupti architecto rerum nisi excepturi sunt quaerat iste quibusdam, non nam possimus ut eveniet.
  </p>
</div>
`;
